transports/janus_websockets.c

```diff
@@ -0,0 +1,906 @@
+		.get_description = janus_websockets_get_description,
+		.get_name = janus_websockets_get_name,
+		.get_author = janus_websockets_get_author,
+		.get_package = janus_websockets_get_package,
+		
+		.send_message = janus_websockets_send_message,
+		.session_created = janus_websockets_session_created,
+		.session_over = janus_websockets_session_over,
+	);
+
+/* Transport creator */
+janus_transport *create(void) {
+	JANUS_LOG(LOG_VERB, "%s created!\n", JANUS_WEBSOCKETS_NAME);
+	return &janus_websockets_transport;
+}
+
+
+/* Useful stuff */
+static gint initialized = 0, stopping = 0;
+static janus_transport_callbacks *gateway = NULL;
+
+/* Logging */
+static int ws_log_level = 0;
+
+/* WebSockets single thread */
+static GThread *thread = NULL;
+void *janus_websockets_thread(void *data);
+
+
+/* WebSocket client session */
+typedef struct janus_websockets_client {
+	struct libwebsocket_context *context;	/* The libwebsock client context */
+	struct libwebsocket *wsi;				/* The libwebsock client instance */
+	GAsyncQueue *messages;					/* Queue of outgoing messages to push */
+	janus_mutex mutex;						/* Mutex to lock/unlock this session */
+	gint session_timeout:1;					/* Whether a Janus session timeout occurred in the core */
+	gint destroy:1;							/* Flag to trigger a lazy session destruction */
+} janus_websockets_client;
+
+
+/* libwebsockets WS context(s) */
+static struct libwebsocket_context *wss = NULL, *swss = NULL,
+	*admin_wss = NULL, *admin_swss = NULL;
+/* libwebsockets sessions that have been closed */
+static GList *old_wss;
+static janus_mutex old_wss_mutex;
+/* Callback for HTTP-related events (automatically rejected) */
+static int janus_websockets_callback_http(struct libwebsocket_context *this, 
+		struct libwebsocket *wsi,
+		enum libwebsocket_callback_reasons reason,
+		void *user, void *in, size_t len);
+/* Callbacks for WebSockets-related events */
+static int janus_websockets_callback(struct libwebsocket_context *this, 
+		struct libwebsocket *wsi,
+		enum libwebsocket_callback_reasons reason,
+		void *user, void *in, size_t len);
+static int janus_websockets_admin_callback(struct libwebsocket_context *this, 
+		struct libwebsocket *wsi,
+		enum libwebsocket_callback_reasons reason,
+		void *user, void *in, size_t len);
+/* Protocol mappings */
+static struct libwebsocket_protocols wss_protocols[] = {
+	{ "http-only", janus_websockets_callback_http, 0, 0 },
+	{ "janus-protocol", janus_websockets_callback, sizeof(janus_websockets_client), 0 },
+	{ NULL, NULL, 0 }
+};
+static struct libwebsocket_protocols admin_wss_protocols[] = {
+	{ "http-only", janus_websockets_callback_http, 0, 0 },
+	{ "janus-admin-protocol", janus_websockets_admin_callback, sizeof(janus_websockets_client), 0 },
+	{ NULL, NULL, 0 }
+};
+/* Helper for debugging reasons */
+const char *janus_websockets_reason_string(enum libwebsocket_callback_reasons reason);
+const char *janus_websockets_reason_string(enum libwebsocket_callback_reasons reason) {
```

Comment:
Apologies, apparently I've missed quite a lot of comments to this pull request... thanks for the heads up, I'll try to make this cleaner in this and other files (I think we have the same thing across different plugins as well).